### PR TITLE
Added following string functions: regex, substr, substring, ltrim, rtrim, trim, upper, lower, concat, concat_ws, length, strcmp

### PR DESCRIPTION
**Pull request by lyndonb-bq**
*Issue #, if available:*
https://github.com/opendistro-for-elasticsearch/sql/issues/710

*Description of changes:*
Added implementation unit testing, manual IT, and doc updates for the following string functions:

- regex
- substr
- substring
- ltrim
- rtrim
- trim
- upper
- lower
- concat
- concat_ws
- length
- strcmp

The following functions have limitations from the specification:

- trim Does not support LEADING/TRAILING/BOTH/FROM commands, only supports string space character trimming
- substr/substring Does not support FROM/TO commands, only comma separated values
- concat/concat_ws Does not support variable arguments being specified, only two strings to concatenate

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+                    STRING, STRING, INTEGER, INTEGER));
+  }
+
+  /**
+   * Gets substring starting at given point, for optional given length.
+   * Form of this function using keywords instead of comma delimited variables is not supported.
+   * Supports following signatures:
+   * (STRING, INTEGER)/(STRING, INTEGER, INTEGER) -> STRING
+   */
+  private FunctionResolver substr() {
```
Comment (penghuo) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java:82]: same with substring?

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        impl(nullMissingHandling((str1, str2) ->
+            new ExprIntegerValue(Integer.compare(
+                str1.stringValue().compareTo(str2.stringValue()), 0))),
+            INTEGER, STRING, STRING));
+  }
+
+  private static ExprValue exprSubstrStart(ExprValue exprValue, ExprValue start) {
+    int startIdx = start.integerValue();
+    if (startIdx == 0) {
+      return new ExprStringValue("");
```
Comment (penghuo) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java:203]: minor, add EMPTY_STRING?

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    String str = exprValue.stringValue();
+    startIdx = (startIdx > 0) ? (startIdx - 1) : (str.length() + startIdx);
+    if (startIdx > str.length()) {
+      return new ExprStringValue("");
+    }
+
+    return new ExprStringValue(str.substring(startIdx));
+  }
+
+  private static ExprValue exprSubstrStartLength(
```
Comment (penghuo) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java:214]: considering merge whith exprSubstrStart

File: core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunctionTest.java [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+    } else {
+      expr = dsl.substr(DSL.literal(strInfo.getExpr()),
+          DSL.literal(strInfo.getStart()),
+          DSL.literal(strInfo.getLen()));
+    }
+    assertEquals(STRING, expr.type());
+    assertEquals(strInfo.getRes(), eval(expr).stringValue());
+  }
+
+  void substringTest(SubstringInfo strInfo) {
```
Comment (penghuo) [inline core/src/test/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunctionTest.java:142]: duplicate with substrTest

**Comment by penghuo** [comment]
Please add IT for PPL also.

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java (hunk 1 above)
Comment (lyndonb-bq) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java:84]: Yeah they are synonyms.

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java (hunk 3 above)
Comment (lyndonb-bq) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java:211]: Merged together to use function below.

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java (hunk 2 above)
Comment (lyndonb-bq) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java]: Added EMPTY_STRING constant.

**Review by rupal-bq — APPROVED** [review]

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+            INTEGER, STRING, STRING));
+  }
+
+  private static ExprValue exprSubstrStart(ExprValue exprValue, ExprValue start) {
+    int startIdx = start.integerValue();
+    if (startIdx == 0) {
+      return new ExprStringValue(EMPTY_STRING);
+    }
+    String str = exprValue.stringValue();
+    return exprSubStr(str, startIdx, 0);
```
Comment (penghuo) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java:202]: the length=0 is misunderstanding, in Java SDK, the definition like below
```
    public String substring(int beginIndex) {
        return substring(beginIndex, length());
    }
```

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java (hunk 1 above)
Comment (penghuo) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/text/TextFunction.java:78]: could you merge them, if they are same.

**Review by penghuo — APPROVED** [review]
Thanks for the change.

**Review by dai-chen — APPROVED** [review]

**Pull request merged**